types/vote_set.go

```diff
@@ -671,6 +672,110 @@ func (voteSet *VoteSet) MakeExtendedCommit(fp FeatureParams) *ExtendedCommit {
 	return ec
 }
 
+// MakeBLSCommit is a variant of MakeExtendedCommit that aggregates BLS signatures.
+//
+// It additionally aggregates the BLS signatures for the block and nil. The
+// resulting commit contains only two aggregated signatures:
+//
+// 1 - aggregated signature for the block (compressed)
+// 2 - aggregated signature for nil (compressed).
+//
+// Note the signatures count is preserved, but only the first signature in the
+// each category (block, nil) is non-empty.
+func (voteSet *VoteSet) MakeBLSCommit() *ExtendedCommit {
+	voteSet.mtx.Lock()
+	defer voteSet.mtx.Unlock()
+
+	if voteSet.signedMsgType != PrecommitType {
+		panic("Cannot MakeBLSCommit() unless VoteSet.Type is PrecommitType")
+	}
+
+	// Make sure we have a 2/3 majority
+	if voteSet.maj23 == nil {
+		panic("Cannot MakeBLSCommit() unless a blockhash has +2/3")
+	}
+
+	// 1. Aggregate the signatures for the block.
+	sigsToAgg := make([][]byte, 0, len(voteSet.votes))
+	for _, v := range voteSet.votes {
+		if v != nil && v.BlockID.IsComplete() {
```

Comment:
Why `v.BlockID.IsComplete()`? Shouldn't it be `!v.BlockID.IsNil()`?
Also, looking at the implementation of `MakeExtendedCommit`, I think we should do the same check here: making sure the block matches `voteSet.maj23` to avoid processing a `voteSet` that, for some reason contains votes for different blocks

Reply:
> Also, you should check that the signature's flag BlockIDFlagis BlockIDFlagCommit.

We don't have a flag at this point.

> Why v.BlockID.IsComplete()?

Copied the logic from `MakeExtendedCommit`.